src/ngraph/runtime/backend.hpp

```diff
@@ -17,83 +17,75 @@
-
-            virtual void enable_performance_data(std::shared_ptr<Function> func, bool enable) {}
-            virtual std::vector<PerformanceCounter>
-                get_performance_data(std::shared_ptr<Function> func) const;
-
-            static bool register_backend(const std::string& name, std::shared_ptr<Backend>);
-
-        protected:
-            void validate_call(std::shared_ptr<const Function> func,
-                               const std::vector<std::shared_ptr<runtime::TensorView>>& outputs,
-                               const std::vector<std::shared_ptr<runtime::TensorView>>& inputs);
-
-        private:
-            static DL_HANDLE open_shared_library(std::string type);
-            static std::map<std::string, std::string> get_registered_device_map();
-            static bool is_backend_name(const std::string& file, std::string& backend_name);
-        };
+        class Backend;
     }
 }
+
+/// @brief Interface to a generic backend.
+///
+/// Backends are responsible for function execution and value allocation.
+class ngraph::runtime::Backend
+{
+public:
+    virtual ~Backend();
+    /// @brief Create a new Backend object
+    /// @param type The name of a registered backend, such as "CPU" or "GPU".
+    ///   To select a subdevice use "GPU:N" where s`N` is the subdevice number.
+    /// @returns shared_ptr to a new Backend or nullptr if the named backend
+    ///   does not exist.
+    static std::shared_ptr<Backend> create(const std::string& type);
+
+    /// @brief Query the list of registered devices
+    /// @returns A vector of all registered devices.
+    static std::vector<std::string> get_registered_devices();
+
+    virtual std::shared_ptr<ngraph::runtime::TensorView>
+        create_tensor(const ngraph::element::Type& element_type, const Shape& shape) = 0;
+
+    /// @brief Return a handle for a tensor for given mem on backend device
+    virtual std::shared_ptr<ngraph::runtime::TensorView> create_tensor(
+        const ngraph::element::Type& element_type, const Shape& shape, void* memory_pointer) = 0;
+
+    template <typename T>
+    std::shared_ptr<ngraph::runtime::TensorView> create_tensor(const Shape& shape)
+    {
+        return create_tensor(element::from<T>(), shape);
+    }
+
+    virtual bool compile(std::shared_ptr<Function> func) = 0;
+
+    virtual bool call(std::shared_ptr<Function> func,
+                      const std::vector<std::shared_ptr<runtime::TensorView>>& outputs,
+                      const std::vector<std::shared_ptr<runtime::TensorView>>& inputs) = 0;
+
+    virtual void remove_compiled_function(std::shared_ptr<Function> func);
+
+    virtual void enable_performance_data(std::shared_ptr<Function> func, bool enable) {}
+    virtual std::vector<PerformanceCounter>
+        get_performance_data(std::shared_ptr<Function> func) const;
+
+    static bool register_backend(const std::string& name, std::shared_ptr<Backend>);
```

Comment:
The backend has two APIs, one for the frameworks and one for the backend implementations. I wanted to keep all of the APIs in backend.hpp rather than split it across Backend and BackendManager. BackendManager is something that I want hidden behind the scene, not something visible to either frameworks or backend implementations. runtime::Backend is really just the API.

Reply:
Spliting the API into two files has some value in my opinion. One is the usage simplicity i.e., you don't need to know which functions are for you - rather all of them can be used by the framework or a specific backend. The other is the decoupling of the dependencies. Which actually is something needs to be address at the nGraph design level but for now I would recommend thinking of splitting this into two parts.

Reply:
I removed the register_backend from Backend and we will use the one in BackendManager. I agree with you on this.